core/src/com/unciv/logic/civilization/CivConstructions.kt

```diff
@@ -0,0 +1,165 @@
+    }
+    
+    fun getFreeBuildings(cityId: String): HashSet<String> {
+        val toReturn = freeBuildings[cityId] ?: hashSetOf()
+        for (city in civInfo.cities) {
+            toReturn.addAll(city.cityConstructions.freeBuildingsProvidedFromThisCity[city.id] ?: hashSetOf())
+        }
+        return toReturn
+    }
+    
+    private fun addFreeBuilding(cityId: String, building: String) {
+        if (!freeBuildings.containsKey(cityId))
+            freeBuildings[cityId] = hashSetOf()
+        freeBuildings[cityId]!!.add(building)
+    }
+    
+    private fun addFreeStatsBuildings() {
+        val statUniquesData = civInfo.getMatchingUniques("Provides the cheapest [] building in your first [] cities for free")
+            .groupBy { it.params[0] }
+            .mapKeys { Stat.valueOf(it.key) }
+            .mapValues { unique -> unique.value.sumOf { it.params[1].toInt() } }
+            .toMutableMap()
+        
+        // Deprecated since 3.16.15
+            statUniquesData[Stat.Culture] = (statUniquesData[Stat.Culture] ?: 0) +
+                civInfo.getMatchingUniques("Immediately creates the cheapest available cultural building in each of your first [] cities for free")
+                    .sumOf { it.params[1].toInt() }
+        //
+        
+        
+        for ((stat, amount) in statUniquesData) {
+            addFreeStatBuildings(stat, amount)
+        }
+    }
+    
+    private fun addFreeStatBuildings(stat: Stat, amount: Int) {
+        for (city in civInfo.cities.take(amount)) {
+            if (freeStatBuildingsProvided[stat]!!.contains(city.id) || !city.cityConstructions.hasBuildableStatBuildings(stat)) continue
+            
+            val builtBuilding = city.cityConstructions.addCheapestBuildableStatBuilding(stat)
+            if (builtBuilding != null) {
+                freeStatBuildingsProvided[stat]!!.add(city.id)
+                addFreeBuilding(city.id, builtBuilding)
+            }
+        }
+    }
+    
+    private fun addFreeSpecificBuildings() {
+        val buildingsUniquesData = (civInfo.getMatchingUniques("Provides a [] in your first [] cities for free")
+            // Deprecated since 3.16.15
+                + civInfo.getMatchingUniques("Immediately creates a [] in each of your first [] cities for free")
+            //
+            ).groupBy { it.params[0] }
+            .mapValues { unique -> unique.value.sumOf { it.params[1].toInt() } }
+
+        for ((building, amount) in buildingsUniquesData) {
+            addFreeBuildings(building, amount)
+        }
+    }
+    
+    private fun addFreeBuildings(building: String, amount: Int) {
+        for (city in civInfo.cities.take(amount)) {
+            if (freeSpecificBuildingsProvided[building]?.contains(city.id) == true || city.cityConstructions.containsBuildingOrEquivalent(building)) continue
+
+            (city.cityConstructions.getConstruction(building) as INonPerpetualConstruction).postBuildEvent(city.cityConstructions)
```

Comment:
This is correct, and this comment isn't aimed specifically at you, but the fact that you're using this here is indicative of a deeper problem in the code.
CityConstructions has an addBuilding function, which is called by the postBuildEvent of the Building.
But, the addBuilding is also called directly from other places.
This is a problem of a couple of sorts - if I, a coder, see the `Cityconstructions.addBuilding` function, my first thought would be that this is what I should call to add a building to the city. There's no way I would think to check the postBuildEvent of the city.
I'm not sure what the best way to solve this would be. The beauty of the postBuildEvent is that it lets the CityConstructions 'build' units and cities in the same way, but from a more holistic perpective this is a detriment to the general codebase...